Replace the custom post-processing shader tutorial (#5946)
godotengine/godot-docs: tutorials/shaders/custom_postprocessing.rst

Introduction
------------

> Godot provides many post-processing effects out of the box, including Bloom, DOF, and SSAO. Sometimes you want to write your own custom effect. Here's how you can do so.

Reviewer: ```suggestion
Godot provides many post-processing effects out of the box, including Bloom, DOF, and SSAO. However, advanced use cases may require custom effects. This article explains how to write your own custom effects.
```

The easiest way to implement a custom post-processing shader is to use Godot's built-in ability to read from
the screen texture. If you're not familiar with this, you should read the :ref:`Screen Reading Shaders
Tutorial <doc_screen-reading_shaders>` first.

Single pass post-processing
---------------------------

> Post-processing effects are shaders applied to a frame after Godot rendered it. To apply a shader to a frame, create a :ref:`CanvasLayer <class_CanvasLayer>`, and give it a :ref:`ColorRect <class_ColorRect>`. Give it a new :ref:`ShaderMaterial <class_ShaderMaterial>`, and set the ``ColorRect``'s layout to "Full Rect".

Reviewer: In the RST docs, single newline characters are ignored, so you can split up paragraphs into multiple lines. Note that this is different from the class reference which does care about single newline characters.

Your scene tree will look something like this:

.. image:: img/post_tree1.png

.. note::

    As of the time of writing, Godot does not support rendering to multiple buffers at the same time. Your
    post-processing shader will not have access to normals or other render passes. You only have
    access to the rendered frame.

> For this demo, we will use this :ref:`Sprite <class_Sprite2D>` of a sheep.

Reviewer: ```suggestion
We will use this image of a sheep as an example.
```

.. image:: img/post_example1.png

> To the ``ColorRect``'s ``ShaderMaterial``, assign a new :ref:`Shader <class_Shader>`. You can access the frame's texture and UV with the built in ``SCREEN_TEXTURE`` and ``SCREEN_UV`` uniforms.

Reviewer: ```suggestion
Assign a new :ref:`Shader <class_Shader>` to the ``ColorRect``'s ``ShaderMaterial``. You can access the frame's texture and UV with the built in ``SCREEN_TEXTURE`` and ``SCREEN_UV`` uniforms.
```

    void fragment()
    {
            vec2 norm_size = size * SCREEN_PIXEL_SIZE;
            bool half = mod(SCREEN_UV.y / 2.0, norm_size.y) / norm_size.y < 0.5;
            vec2 uv = SCREEN_UV + vec2(norm_size.x * 0.5 * float(half), 0.0);
            vec2 center_uv = floor(uv / norm_size) * norm_size;
            vec2 norm_uv = mod(uv, norm_size) / norm_size;
            center_uv += mix(vec2(0.0, 0.0),
                             mix(mix(vec2(norm_size.x, -norm_size.y),
                                     vec2(0.0, -norm_size.y),
                                     float(norm_uv.x < 0.5)),
                                 mix(vec2(0.0, -norm_size.y),
                                     vec2(-norm_size.x, -norm_size.y),
                                     float(norm_uv.x < 0.5)),
                                 float(half)),
                             float(norm_uv.y < 0.3333333) * float(norm_uv.y / 0.3333333 < (abs(norm_uv.x - 0.5) * 2.0)));

            COLOR = textureLod(SCREEN_TEXTURE, center_uv, 0.0);
    }

The sheep will look something like this:

.. image:: img/post_example2.png

Multi-pass post-processing
--------------------------

> Some post-processing effects like blurs are resource intensive. But you can make them run a lot faster if you break them down in multiple passes. In a multipass material, each pass takes the result from the previous pass as an input and processes it.

Reviewer: ```suggestion
Some post-processing effects like blurs are resource intensive. You can make them run a lot faster if you break them down in multiple passes. In a multipass material, each pass takes the result from the previous pass as an input and processes it.
```
